libmve/mveplay.cpp

```diff
@@ -0,0 +1,481 @@
+    count = get_short(data + 4);
+  } else {
+    count = 1;
+  }
+
+  if (minor > 1) {
+    truecolor = get_short(data + 6);
+  } else {
+    truecolor = 0;
+  }
+
+  g_width = w << 3;
+  g_height = h << 3;
+
+  /* TODO: * 4 causes crashes on some files */
+  /* only malloc once */
+  if (g_vBuffers == nullptr)
+    g_vBackBuf1 = g_vBuffers = mve_alloc(g_width * g_height * 8);
+  if (truecolor) {
+    g_vBackBuf2 = (unsigned short *)g_vBackBuf1 + (g_width * g_height);
+  } else {
+    g_vBackBuf2 = (unsigned char *)g_vBackBuf1 + (g_width * g_height);
+  }
+
+  memset(g_vBackBuf1, 0, g_width * g_height * 4);
+
+  g_truecolor = truecolor;
+
+  return 1;
+}
+
+static int display_video_handler(unsigned char major, unsigned char minor, unsigned char *data, int len,
+                                 void *context) {
+  if (g_destX == -1) // center it
+    g_destX = (g_screenWidth - g_width) >> 1;
+  if (g_destY == -1) // center it
+    g_destY = (g_screenHeight - g_height) >> 1;
+
+  mve_showframe((unsigned char *)g_vBackBuf1, g_width, g_height, 0, 0, g_width, g_height, g_destX, g_destY,
+                g_truecolor);
+
+  g_frameUpdated = 1;
+
+  return 1;
+}
+
+static int init_video_handler(unsigned char major, unsigned char minor, unsigned char *data, int len, void *context) {
+  short width, height;
+
+  if (video_initialized)
+    return 1; /* maybe we actually need to change width/height here? */
+  else
+    video_initialized = 1;
+
+  width = get_short(data);
+  height = get_short(data + 2);
+  g_screenWidth = width;
+  g_screenHeight = height;
+
+  return 1;
+}
+
+static int video_palette_handler(unsigned char major, unsigned char minor, unsigned char *data, int len,
+                                 void *context) {
+  short start, count;
+  unsigned char *p;
+
+  start = get_short(data);
+  count = get_short(data + 2);
+
+  p = data + 4;
+
+  mve_setpalette(p - 3 * start, start, count);
+
+  return 1;
+}
+
+static int video_codemap_handler(unsigned char major, unsigned char minor, unsigned char *data, int len,
+                                 void *context) {
+  g_pCurMap = data;
+  g_nMapLength = len;
+  return 1;
+}
+
+static int video_data_handler(unsigned char major, unsigned char minor, unsigned char *data, int len, void *context) {
+  unsigned char *temp;
+
+  short nFrameHot = get_short(data);
+  short nFrameCold = get_short(data + 2);
+  short nXoffset = get_short(data + 4);
+  short nYoffset = get_short(data + 6);
+  short nXsize = get_short(data + 8);
+  short nYsize = get_short(data + 10);
+  unsigned short nFlags = get_ushort(data + 12);
+
+  if (nFlags & 1) {
+    temp = (unsigned char *)g_vBackBuf1;
+    g_vBackBuf1 = g_vBackBuf2;
+    g_vBackBuf2 = temp;
+  }
+
+  /* convert the frame */
+  if (g_truecolor) {
+    decodeFrame16((unsigned char *)g_vBackBuf1, g_pCurMap, g_nMapLength, data + 14, len - 14);
+  } else {
+    decodeFrame8((unsigned char *)g_vBackBuf1, g_pCurMap, g_nMapLength, data + 14, len - 14);
```

Comment:
I know this code is pulled from D2X but do we actually need 8bit decoding instead of 16?

Reply:
8bit decoding is still used for Dolby Video.